external/js/cothority/src/byzcoin/proof.ts

```diff
@@ -446,6 +417,53 @@ class InclusionProof extends Message<InclusionProof> {
 
         return h.digest();
     }
+
+    /**
+     * Check if the key exists in the proof
+     *
+     * @returns true when it exists, false otherwise
+     * @throws for corrupted proofs
+     */
+    exists(key: Buffer): boolean {
+        if (key.length === 0) {
+            throw new Error("key is nil");
+        }
+        if (this.interiors.length === 0) {
+            throw new Error("no interior node");
+        }
+
+        const bits = hashToBits(key);
+        let expectedHash = this.hashInterior(0);
+
+        let i = 0;
+        for (; i < this.interiors.length; i++) {
+            if (!expectedHash.equals(this.hashInterior(i))) {
+                throw new Error("invalid interior node");
+            }
+
+            if (bits[i]) {
+                expectedHash = this.interiors[i].left;
+            } else {
+                expectedHash = this.interiors[i].right;
+            }
+        }
+
+        if (expectedHash.equals(this.hashLeaf())) {
+            if (_.difference(bits.slice(0, i), this.leaf.prefix).length !== 0) {
+                throw new Error("invalid prefix in leaf node");
+            }
+
+            return this.key.equals(key);
+        } else if (expectedHash.equals(this.hashEmpty())) {
+            if (_.difference(bits.slice(0, i), this.empty.prefix).length !== 0) {
+                throw new Error("invalid prefix in empty node");
+            }
+
+            return false;
+        }
```

Comment:
Can't you merge both blocks?

Reply:
that's a copy of the code in go - looks more complicated to merge them than to let it be like that.